### PR TITLE
Add Moment Morph Interpolation

### PR DESCRIPTION
**Pull request by RuneDominik**
_This is a new version of #221 that gets rid of most classes. It thus supersedes #221. @maxnoe's comments in #221 were included in this version_

This PR introduces another interpolation method (Moment Morphing, see [1]). While performing comparably to the already existing Quantile Interpolation the main points in favor of this method is that it can be can be separated into two steps. First, one has to compute interpolation coefficients and then these are used to morph and interpolate the input pdfs. When changing the coefficients (e.g. changing interpolation to extrapolation or changing from interpolation in a triangle to interpolation in a rectangle) the second part remains unchanged. Consequently, using this for extrapolation is quite simple.

This PR is quite extensive in its additions. Most of these are actually tests, checking all kind of broadcasting and border cases.

[1] M. Baak, S. Gadatsch, R. Harrington and W. Verkerke (2015). Interpolation between multi-dimensional histograms using a new non-linear moment morphing method. Nucl. Instrum. Methods Phys. Res. A 771, 39-48. https://doi.org/10.1016/j.nima.2014.10.033

### CODE REVIEW

File: pyirf/interpolation/moment_morph_interpolator.py [hunk 1]
```diff
@@ -0,0 +1,354 @@
...
+    """
+    # Create a flattend version of self.bin_contents to ease broadcasting
+    intermediate_bin_contentents = bin_contents.reshape(-1, bin_contents.shape[-1])
+
+    # Find the bin where each point x is located in
+    binnr = np.digitize(x, bin_edges).reshape(-1, bin_contents.shape[-1])
+
+    # np.digitize returns 0 if x is below the first and len(bins) if x is above the last bin,
+    # set these to 0 in a copy of the binnr to avoid errors later
+    binnr_copy = np.copy(binnr)
```
Comment (maxnoe) [inline pyirf/interpolation/moment_morph_interpolator.py:79]: why copy?

It's not an input, so ok to mutate. also, where should create a fresh array again.

File: pyirf/interpolation/moment_morph_interpolator.py [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+    binnr_copy = np.where((binnr == 0) | (binnr == len(bin_edges)), 0, binnr_copy)
```
Comment (maxnoe) [inline pyirf/interpolation/moment_morph_interpolator.py:80]: Setting binnr = 0 where it is 0 already seems superfluous

File: pyirf/interpolation/moment_morph_interpolator.py [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+    -------
+    y: numpy.ndarray, shape=(N, ..., M)
+        Array of the bin-heights at the M points x, set to 0 at each point outside the histogram
+
+    """
+    # Create a flattend version of self.bin_contents to ease broadcasting
+    intermediate_bin_contentents = bin_contents.reshape(-1, bin_contents.shape[-1])
+
+    # Find the bin where each point x is located in
+    binnr = np.digitize(x, bin_edges).reshape(-1, bin_contents.shape[-1])
```
Comment (maxnoe) [inline pyirf/interpolation/moment_morph_interpolator.py:75]: There is also `pyirf.binning.calculate_bin_indices` which does most (maybe all?) of this digitize checking magic...

File: pyirf/interpolation/moment_morph_interpolator.py [hunk 4]
```diff
@@ -0,0 +1,353 @@
...
+    # Set zeroth grid point as reference value
+    m0 = grid_points[0, :]
+
+    # Compute matrix M as in eq. (2) of [1]
+    j = np.arange(0, grid_points.shape[0], 1)
+    m_ij = (grid_points - m0) ** j
+
+    # Compute coefficients, eq. (6) from [1]
+    return np.einsum(
+        "...j, ji -> ...i", ((target_point - m0) ** j), np.linalg.inv(m_ij)
```
Comment (maxnoe) [inline pyirf/interpolation/moment_morph_interpolator.py:128]: superfluous parentheses around the second argument

File: pyirf/interpolation/moment_morph_interpolator.py [hunk 5]
```diff
@@ -0,0 +1,353 @@
...
+        References
+        ----------
+        .. [1] M. Baak, S. Gadatsch, R. Harrington and W. Verkerke (2015). Interpolation between
+               multi-dimensional histograms using a new non-linear moment morphing method
+               Nucl. Instrum. Methods Phys. Res. A 771, 39-48. https://doi.org/10.1016/j.nima.2014.10.033
+        """
+        if self.grid_dim == 1:
+            interpolant = self._interpolate1D(target_point)
+        elif self.grid_dim == 2:
+            interpolant = self._interpolate2D(target_point)
```
Comment (maxnoe) [inline pyirf/interpolation/moment_morph_interpolator.py:351]: `else: raise...` ?

Comment (maxnoe) [reply pyirf/interpolation/moment_morph_interpolator.py:351]: Nevermind, checked in `__init__`

File: pyirf/interpolation/tests/test_moment_morph_interpolator.py [hunk 6]
```diff
@@ -0,0 +1,872 @@
...
+    }
+
+
+@pytest.fixture
+def simple_2D_data(bins):
+    grid = np.array([[20, 20], [60, 20], [40, 60]])
+    target = np.array([25, 25])
+    bin_contents = np.array(
+        [
+            np.diff(
```
Comment (maxnoe) [inline pyirf/interpolation/tests/test_moment_morph_interpolator.py:52]: There is a lot of repetition here, seems to be easy to factor out to something like

```
def _binned_pdf(dist, bins):
     return np.diff(dist.cdf(bins))
```
?

**Review by maxnoe — CHANGES_REQUESTED** [review]
This looks good aside from readability in the tests, would be great if the test case data would be easier to read, e.g. by factoring out some functions to reduce the nesting levels.

File: pyirf/interpolation/tests/test_moment_morph_interpolator.py (hunk 6 above)
Comment (RuneDominik) [reply pyirf/interpolation/tests/test_moment_morph_interpolator.py]: Yeah, thats right. Doing so actually got rid of some 250 lines of code :sweat_smile:

**Review by maxnoe — APPROVED** [review]

**Pull request merged**